org.civicrm.reqangular/src/common/components/heatmap.component.js

```diff
@@ -0,0 +1,131 @@
+/* eslint-env amd */
+
+define([
+  'common/lodash',
+  'common/moment',
+  'common/modules/components'
+], function (_, moment, components) {
+  return components.component('heatmap', {
+    bindings: {
+      values: '<'
+    },
+    controller: heatmapController,
+    controllerAs: 'heatmap',
+    template: ['$templateCache', function ($templateCache) {
+      return $templateCache.get('components/heatmap.html');
+    }]
+  });
+
+  function heatmapController () {
+    var vm = this;
+
+    vm.heatmapDays = [];
+
+    vm.$onChanges = $onChanges;
+
+    /**
+     * Implementes the $onChanges method for angular controllers. When the
+     * values binding is ready, it maps the values to the heatmap.
+     */
+    function $onChanges (changes) {
+      if (changes.values) {
+        mapValuesToHeatmapDays();
+      }
+    }
+
+    /**
+     * Returns a clean copy of the heatmap values that are not false.
+     *
+     * @return {Array}
+     */
+    function getCleanHeatValues () {
+      return Object.values(vm.values).filter(function (value) {
+        return value !== false;
+      });
+    }
+
+    /**
+     * Given a range of heat levels and a value, it will return the closest
+     * heat level for the value.
+     *
+     * @param {Object[]} heatLevelRanges - An array of heat levels and their values.
+     * @param {Number} heatValue - The value to compare against heat levels.
+     *
+     * @param {String}
+     */
+    function getHeatLevelForValue (heatLevelRanges, heatValue) {
+      var heatRange = heatLevelRanges.reduce(function (previous, current) {
+        return Math.abs(current.value - heatValue) < Math.abs(previous.value - heatValue)
```

Comment:
Put `current.value - heatValue` and `previous.value - heatValue` in two variables with descriptive names, so the comparison is more comprehensible